### PR TITLE
Fix/tip reuse with multi channeling try2

### PR DESCRIPTION
**Pull request by m-i-sadowski**
This PR contains a refactor of the liquid handling instruction hierarchy to 
change the TransferInstruction class to permit it to define multiple 
multichannel transfers where previously it could define at most one. 

This has the consequence that a single transfer can now generate a block
of multichannel instructions instead of only one. Since tip changes are always
generated at the start of a new block this was causing excessive tip usage. 
Consequently this refactor should allow for considerably more conservative
tip use.

### CODE REVIEW

**Comment by m-i-sadowski** [comment]
Still some tests to do but functionally this looks good so far

File: microArch/driver/liquidhandling/transferblock.go [hunk 1]
```diff
@@ -79,14 +79,35 @@ func (ti TransferBlockInstruction) Generate(ctx context.Context, policy *wtype.L
...
+		r := make([]RobotInstruction, len(inss))
+
+		for i, ins := range inss {
+			r[i] = RobotInstruction(ins)
+		}
+
+		return r
+	}
+
+	inss = fromTransfers(mergeTransfers(toTransfers(inss), policy))
```
Comment (haydnKing) [inline microArch/driver/liquidhandling/transferblock.go]: Did we establish whether mergeTransfers is now working as expected? Previously it was failing to merge successive transfers (which allowed the planning numerical instability to end up causing the multichannel non-determinism)

File: microArch/driver/liquidhandling/transferblock.go [hunk 2]
```diff
@@ -485,69 +505,51 @@ func (ti TransferBlockInstruction) GetParameter(p string) interface{} {
...
+	// merge only if the merge doesn't break either
 
-	// we just compare component names
+	ins3 := ins.Dup()
+	ins3.MergeWith(ins2)
 
-	m := a2Map(ins1.Components)
+	m1 := GetPolicyFor(policy, ins)["CAN_MULTI"].(bool)
+	m2 := GetPolicyFor(policy, ins2)["CAN_MULTI"].(bool)
+	m3 := GetPolicyFor(policy, ins3)["CAN_MULTI"].(bool)
```
Comment (haydnKing) [inline microArch/driver/liquidhandling/transferblock.go]: There's some discussion about allowing elements to set LHPolicy items in the future, so we might want to think about error checking these, e.g. through implementing some accessors like `LHPolicy.getBool(name string) bool, err`

File: microArch/scheduler/liquidhandling/executionplanner3.go [hunk 3]
```diff
@@ -52,8 +53,11 @@ func ExecutionPlanner3(ctx context.Context, request *LHRequest, robot *liquidhan
 
 			tfb := liquidhandling.NewTransferBlockInstruction(ch.Values)
 
+			fmt.Println("DEPTH ", depth, " LEN TFB: ", len(ch.Values))
```
Comment (haydnKing) [inline microArch/scheduler/liquidhandling/executionplanner3.go]: Not sure if this is useful for users or leftover from debugging/development ;)

File: microArch/driver/liquidhandling/transferblock.go (hunk 2 above)
Comment (m-i-sadowski) [reply microArch/driver/liquidhandling/transferblock.go]: that would be safer, however the plan is to validate setting these items at source and only permit the correct datatype to be used. See aparam.go in wtype.

File: microArch/driver/liquidhandling/transferblock.go (hunk 1 above)
Comment (m-i-sadowski) [reply microArch/driver/liquidhandling/transferblock.go]: This is a total rewrite of how transfer merging happens. It should be considerably more robust. So far my tests have found it to perform better but it's worth trying against that same test case to confirm, I'll do that.

**Comment by m-i-sadowski** [comment]
Tests pass locally.

File: microArch/driver/liquidhandling/transferblock.go [hunk 4]
```diff
@@ -485,69 +505,51 @@ func (ti TransferBlockInstruction) GetParameter(p string) interface{} {
...
 
-	if len(m) > 1 {
-		return false
+	if !xor(m1, m2) {
```
Comment (ddn0) [inline microArch/driver/liquidhandling/transferblock.go]:
```
if m1 == m2 {
  return m1 == m3
}
```

?

File: microArch/driver/liquidhandling/transferinstruction.go [hunk 5]
```diff
@@ -273,45 +228,47 @@ func plateTypeArray(ctx context.Context, types []string) ([]*wtype.LHPlate, erro
...
 
-	// the TransferBlock instruction takes into account the destinations being OK
-	// splits instructions into potentially multiable blocks on that basis
-	// and finds sources for them
+	return r
+}
 
-	// -- the transfer block ensures these instructions are at most multi long
-	// all of the below assumes we can't span multiple plates
+func (ins *TransferInstruction) testParallelSet(ctx context.Context, channel *wtype.LHChannelParameter, which int) bool {
```
Comment (ddn0) [inline microArch/driver/liquidhandling/transferinstruction.go]: rename to `canParallelSet`? or `validateParallelSet`?

File: microArch/scheduler/liquidhandling/executionplanner3.go [hunk 6]
```diff
@@ -54,6 +55,7 @@ func ExecutionPlanner3(ctx context.Context, request *LHRequest, robot *liquidhan
 
 			request.InstructionSet.Add(tfb)
 		}
+		depth += 1
```
Comment (ddn0) [inline microArch/scheduler/liquidhandling/executionplanner3.go]: does anyone use this?

Comment (m-i-sadowski) [reply microArch/scheduler/liquidhandling/executionplanner3.go]: Leftover debug... removed

File: microArch/driver/liquidhandling/transferblock.go (hunk 4 above)
Comment (m-i-sadowski) [reply microArch/driver/liquidhandling/transferblock.go]: Doh! Yes. That's simpler. Much much simpler. Face, meet palm.

File: microArch/driver/liquidhandling/compositerobotinstruction.go [hunk 7]
```diff
@@ -387,17 +310,16 @@ func NewMultiChannelBlockInstruction() *MultiChannelBlockInstruction {
...
-	ins.PltFrom = append(ins.PltFrom, mct.PltFrom)
-	ins.PltTo = append(ins.PltTo, mct.PltTo)
-	ins.WellFrom = append(ins.WellFrom, mct.WellFrom)
-	ins.WellTo = append(ins.WellTo, mct.WellTo)
-	ins.Volume = append(ins.Volume, mct.Volume)
-	ins.FPlateType = append(ins.FPlateType, mct.FPlateType)
-	ins.TPlateType = append(ins.TPlateType, mct.TPlateType)
-	ins.FVolume = append(ins.FVolume, mct.FVolume)
-	ins.TVolume = append(ins.TVolume, mct.TVolume)
-	ins.Prms = mct.Channel
```
[the remark this answers is not in the document]
Comment (m-i-sadowski) [reply microArch/driver/liquidhandling/compositerobotinstruction.go]: Yes

File: microArch/driver/liquidhandling/transferinstruction.go (hunk 5 above)
Comment (m-i-sadowski) [reply microArch/driver/liquidhandling/transferinstruction.go]: done

**Pull request merged**